go/pkg/cas/upload.go

```diff
@@ -662,6 +677,160 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
 	})
 }
 
+// stream uploads the item using ByteStream service.
+//
+// If the blob is already uploaded, then the function returns quickly and
+// without an error.
+func (u *uploader) stream(ctx context.Context, item *uploadItem, updateCacheStats bool) error {
+	if err := u.semByteStreamWrite.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semByteStreamWrite.Release(1)
+
+	// Prepare compressor, if needed.
+	enableCompression := item.Digest.SizeBytes >= u.CompressedBytestreamThreshold
+	var enc *zstd.Encoder
+	if enableCompression {
+		enc = zstdEncoders.Get().(*zstd.Encoder)
+		defer func() {
+			enc.Close()
+			zstdEncoders.Put(enc)
+		}()
+	}
+
+	streamBuf := u.streamBufs.Get().([]byte)
```

Comment:
Move this to `streamFromReader`? I don't expect retries to happen often enough to actually impact performance.

Reply:
done